### PR TITLE
Fix bgupdate error if index already exists

### PR DESCRIPTION
**Pull request by richvdh**
When creating a new table index in the background, only do so if it does not already exist. See code comments for why. Fixes https://github.com/matrix-org/synapse/issues/2135.

Also, make sure we restore the autocommit flag when we're done, otherwise we get more failures from other operations later on. Fixes https://github.com/matrix-org/synapse/issues/1890 (hopefully).

### CODE REVIEW

File: synapse/storage/background_updates.py [hunk 1]
```diff
@@ -228,46 +228,74 @@ def register_background_index_update(self, update_name, index_name,
...
-        def create_index(conn):
+            try:
+                c = conn.cursor()
+
+                # CREATE INDEX IF NOT EXISTS was only added in postgres 9.5 :/.
+                #
+                # instead see if there's a record in pg_class for it.
+                #
+                c.execute("SELECT 1 from pg_class where relname = %s",
+                          (index_name,))
```
Comment (erikjohnston) [inline synapse/storage/background_updates.py]: You'll want to be careful here and check the schema as well

**Comment by erikjohnston** [comment]
LGTM, I wonder why the commit tests are failing?

**Comment by richvdh** [comment]
> LGTM, I wonder why the commit tests are failing?

Because I didn't branch from latest develop, I think

**Comment by erikjohnston** [comment]
Ah, fair

**Pull request merged**